File: web/lib/llm/generate.ts
/**
 * generate.ts — the generate -> validate -> repair loop.
 *
 * One retry, not more. Two reasons: (1) against Gemini's free-tier RPM
 * caps and a still-possible metered provider behind it, unbounded
 * retries are a real cost, not just latency; (2) if a well-specified
 * prompt still fails validation twice, the problem is almost always the
 * prompt or the validator, not transient noise — a third blind attempt
 * rarely helps. Callers get a clear error on failure and decide their
 * own fallback (e.g. brief-builder.ts's grounded, non-fabricating path).
 */
import type { LLMProvider } from "./types";

export class GenerationValidationError extends Error {}

export async function generateValidated<T>(
  provider: LLMProvider,
  prompt: string,
  validate: (raw: string) => T,
  opts?: { system?: string }
): Promise<T> {
  const firstAttempt = await provider.generateText(prompt, opts);
  try {
    return validate(firstAttempt);
  } catch (firstError) {
    const repairPrompt = `${prompt}\n\nYour previous response failed validation with this error:\n${
      firstError instanceof Error ? firstError.message : String(firstError)
    }\n\nYour previous response was:\n${firstAttempt.slice(0, 1000)}\n\nFix the issue and respond again, following the required format exactly.`;

    const secondAttempt = await provider.generateText(repairPrompt, opts);
    try {
      return validate(secondAttempt);
    } catch (secondError) {
      throw new GenerationValidationError(
        `Generation failed validation twice. Last error: ${
          secondError instanceof Error ? secondError.message : String(secondError)
        }`
      );
    }
  }
}

File: src/pipeline/llm/generate.py
"""
generate.py — generate -> validate -> repair loop, TS mirror in
web/lib/llm/generate.ts. One retry, not more — see that file's doc
comment for why (free-tier request budgets, and a prompt/validator
problem rarely resolves on a third blind attempt).
"""

from __future__ import annotations

from typing import Callable, TypeVar

T = TypeVar("T")


class GenerationValidationError(RuntimeError):
    """Raised when generation fails validation on both the original and repair attempt."""


def generate_validated(
    call_fn: Callable[[str], str],
    prompt: str,
    validate_fn: Callable[[str], T],
) -> T:
    """
    call_fn: takes a prompt string, returns raw provider text (e.g.
        functools.partial(call_gemini, api_key=..., system=...))
    validate_fn: parses/validates raw text, returns the validated result,
        or raises on failure.
    """
    first_attempt = call_fn(prompt)
    try:
        return validate_fn(first_attempt)
    except Exception as first_error:  # noqa: BLE001 - validator can raise anything
        repair_prompt = (
            f"{prompt}\n\n"
            f"Your previous response failed validation with this error:\n{first_error}\n\n"
            f"Your previous response was:\n{first_attempt[:1000]}\n\n"
            "Fix the issue and respond again, following the required format exactly."
        )
        second_attempt = call_fn(repair_prompt)
        try:
            return validate_fn(second_attempt)
        except Exception as second_error:  # noqa: BLE001
            raise GenerationValidationError(
                f"Generation failed validation twice. Last error: {second_error}"
            ) from second_error
